### PR TITLE
Distributions types

### PR DESCRIPTION
**Pull request by pannga**

### CODE REVIEW

File: packages/sdk/src/types/Distributions.ts [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+	order?: 'id' | 'voucher_code' | 'tracking_id' | 'customer_id' | 'created_at' | 'channel'
+	campaign?: string
+	customer?: string
+	voucher?: string
+	result?: string
+	voucher_type?: string
+	is_referral_code?: boolean
+	filters?: {
+		junction?: string
+		[filter_condition: string]: any
```
Comment (eddyw) [inline packages/sdk/src/types/Distributions.ts:101]: Use `[filter_condition: string]: $FixMe` type instead, so we can come back to it later to fix it. 

The `$FixMe` type is an alias for `any` ( `type $FixMe = any` ), the difference is that `$FixMe` implies `any` is used but needs to be fixed later, but explicitly specifying `any` implies the change was intentional and doesn't need to be fixed.

Let's use that as a convention (I should have added a `Conventions` section to `Contributing.md` 😅 )

Comment (pannga) [reply packages/sdk/src/types/Distributions.ts:101]: Will do, thanks!

File: packages/sdk/src/types/Distributions.ts [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+	type?: 'PERCENT'
+	percent_off?: number
+}
+
+interface DistributionsPublicationsVoucher {
+	code?: string
+	object?: 'voucher'
+	campaign?: string
+	gift?: {
+		amount?: number
```
Comment (darekg11) [inline packages/sdk/src/types/Distributions.ts:20]: It's always available if `gift` exists so I would drop the `?`

File: packages/sdk/src/types/Distributions.ts [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+		balance?: number
```
Comment (darekg11) [inline packages/sdk/src/types/Distributions.ts:21]: It's always available if `gift` exists so I would drop the `?`

File: packages/sdk/src/types/Distributions.ts [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+	}
+	loyalty_card?: {
+		amount?: number
+		balance?: number
+	}
+	is_referral_code?: boolean
+	discount?:
```
Comment (darekg11) [inline packages/sdk/src/types/Distributions.ts:28]: I would just go with:
```
{
type: "UNIT" | "AMOUNT" | "DISCOUNT",
unit_off?: number
effect?: string
amount_off?: number
percent_off?: number
amount_limit?: number
}
```

File: packages/sdk/src/types/Distributions.ts [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+	unit_off?: number
+	effect?: string
+}
+interface DistributionsPublicationsVoucherDiscountAmount {
+	type?: 'AMOUNT'
+	amount_off?: number
+}
+interface DistributionsPublicationsVoucherDiscountPercent {
+	type?: 'PERCENT'
+	percent_off?: number
```
Comment (darekg11) [inline packages/sdk/src/types/Distributions.ts:12]: It can also has: `amount_limit` property

File: packages/sdk/src/types/Distributions.ts [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+interface DistributionsPublicationsVoucher {
+	code?: string
+	object?: 'voucher'
+	campaign?: string
+	gift?: {
+		amount?: number
+		balance?: number
+	}
+	loyalty_card?: {
+		amount?: number
```
Comment (darekg11) [inline packages/sdk/src/types/Distributions.ts:24]: It's always available if `loyalty_card` exists so I would drop the `?`

File: packages/sdk/src/types/Distributions.ts [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+		balance?: number
```
Comment (darekg11) [inline packages/sdk/src/types/Distributions.ts:25]: It's always available if `loyalty_card` exists so I would drop the `?`

File: packages/sdk/src/types/Distributions.ts [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+}
+
+interface DistributionsPublicationsCreateResponseVoucher {
+	id?: string
+	code?: string
+	campaign?: string
+	campaign_id?: string
+	category?: string
+	type?: string
+	discount?:
```
Comment (darekg11) [inline packages/sdk/src/types/Distributions.ts:41]: I would just go with:
```
{
type: "UNIT" | "AMOUNT" | "DISCOUNT",
unit_off?: number
effect?: string
amount_off?: number
percent_off?: number
amount_limit?: number
}
```

File: packages/sdk/src/types/Distributions.ts [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+		| DistributionsPublicationsVoucherDiscountUnit
+		| DistributionsPublicationsVoucherDiscountAmount
+		| DistributionsPublicationsVoucherDiscountPercent
+	gift?: {
+		amount?: number
```
Comment (darekg11) [inline packages/sdk/src/types/Distributions.ts:46]: It's always available if `gift` exists so I would drop the `?`

File: packages/sdk/src/types/Distributions.ts [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+		balance?: number
```
Comment (darekg11) [inline packages/sdk/src/types/Distributions.ts:47]: It's always available if `gift` exists so I would drop the `?`

File: packages/sdk/src/types/Distributions.ts [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+	}
+	loyalty_card?: {
+		amount?: number
```
Comment (darekg11) [inline packages/sdk/src/types/Distributions.ts:50]: It's always available if `loyalty_card` exists so I would drop the `?`

File: packages/sdk/src/types/Distributions.ts [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+		balance?: number
```
Comment (darekg11) [inline packages/sdk/src/types/Distributions.ts:51]: It's always available if `loyalty_card` exists so I would drop the `?`

File: packages/sdk/src/types/Distributions.ts [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+	}
+	start_date?: string
+	expiration_date?: string
+	validity_timeframe?: string
+	validity_day_of_week?: string
+	active?: boolean
+	additional_info?: string
+	metadata?: Record<string, any>
+	assets?: {
```
Comment (darekg11) [inline packages/sdk/src/types/Distributions.ts:60]: Always there - would drop `?`

File: packages/sdk/src/types/Distributions.ts [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+		qr?: {
```
Comment (darekg11) [inline packages/sdk/src/types/Distributions.ts:61]: Drop `?`

File: packages/sdk/src/types/Distributions.ts [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+			id?: string
```
Comment (darekg11) [inline packages/sdk/src/types/Distributions.ts:62]: Drop `?`

File: packages/sdk/src/types/Distributions.ts [hunk 16]
```diff
@@ -0,0 +1,168 @@
...
+			url?: string
```
Comment (darekg11) [inline packages/sdk/src/types/Distributions.ts:63]: Drop `?`

File: packages/sdk/src/types/Distributions.ts [hunk 17]
```diff
@@ -0,0 +1,168 @@
...
+		}
+		barcode?: {
```
Comment (darekg11) [inline packages/sdk/src/types/Distributions.ts:65]: Drop `?`

File: packages/sdk/src/types/Distributions.ts [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+			id?: string
```
Comment (darekg11) [inline packages/sdk/src/types/Distributions.ts:66]: Drop `?`

File: packages/sdk/src/types/Distributions.ts [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+			url?: string
```
Comment (darekg11) [inline packages/sdk/src/types/Distributions.ts:67]: Drop `?`

File: packages/sdk/src/types/Distributions.ts [hunk 20]
```diff
@@ -0,0 +1,168 @@
...
+		}
+	}
+	is_referral_code?: boolean
+	created_at?: string
```
Comment (darekg11) [inline packages/sdk/src/types/Distributions.ts:71]: Drop `?`

File: packages/sdk/src/types/Distributions.ts [hunk 21]
```diff
@@ -0,0 +1,168 @@
...
+	updated_at?: string
+	holder_id?: string
+	object?: string
```
Comment (darekg11) [inline packages/sdk/src/types/Distributions.ts:74]: Drop `?` and it's always `voucher`

File: packages/sdk/src/types/Distributions.ts [hunk 22]
```diff
@@ -0,0 +1,168 @@
...
+	publish?: {
```
Comment (darekg11) [inline packages/sdk/src/types/Distributions.ts:75]: I think we could drop `?` and the same on every property inside.

File: packages/sdk/src/types/Distributions.ts [hunk 23]
```diff
@@ -0,0 +1,168 @@
...
+		object?: string
+		count?: number
+		url?: string
+	}
+	redemption?: {
```
Comment (darekg11) [inline packages/sdk/src/types/Distributions.ts:80]: I think we could drop `?` and the same on every property inside.

File: packages/sdk/src/types/Distributions.ts [hunk 24]
```diff
@@ -0,0 +1,168 @@
...
+	order?: 'id' | 'voucher_code' | 'tracking_id' | 'customer_id' | 'created_at' | 'channel'
+	campaign?: string
+	customer?: string
+	voucher?: string
+	result?: string
+	voucher_type?: string
+	is_referral_code?: boolean
+	filters?: {
+		junction?: string
+		[filter_condition: string]: $FixMe
```
[the remark this answers is not in the document]
Comment (pannga) [reply packages/sdk/src/types/Distributions.ts:101]: I would need to know every possible filter_condition we could use here.

`filters[<field_name>][conditions][<operator>][<index>]=<value>`
so here, 'filter_condition' should be replaced with possible `field_names`.

`filters?: {
		junction?: 'OR' | 'AND'
		[filter_condition: 'customer_id' | 'other_condition_1' | 'other_condition_2']: {
                      conditions: {
                         [condition: '$in' | '$not' | 'etc']: string[]
                                }
                         }
	}`

But I don't know the full spec of this filters objects

[the remark this answers is not in the document]
Comment (darekg11) [reply packages/sdk/src/types/Distributions.ts:108]: Let's do it at the end when everything is merged.

File: packages/sdk/src/types/Distributions.ts [hunk 25]
```diff
@@ -0,0 +1,161 @@
...
+	is_referral_code?: boolean
+	discount?: VoucherDiscount
+}
+
+interface DistributionsPublicationsCreateResponseVoucher {
+	id: string
+	code: string
+	campaign?: string
+	campaign_id?: string
+	category?: VoucherType
```
Comment (darekg11) [inline packages/sdk/src/types/Distributions.ts:49]: Category should be string and type should be `VoucherType`

File: packages/sdk/src/types/helpers.ts [hunk 26]
```diff
@@ -0,0 +1,2 @@
+
+export interface CampaignType:
```
Comment (darekg11) [inline packages/sdk/src/types/helpers.ts:2]: ??

File: packages/sdk/src/types/Distributions.ts [hunk 27]
```diff
@@ -0,0 +1,161 @@
...
+	voucher_type?: 'GIFT' | 'DISCOUNT' | 'LOYALTY_CARD' | 'LUCKY_DRAW'
+	is_referral_code?: boolean
+	filters?: {
+		junction?: 'OR' | 'AND'
+		[filter_condition: string]: $FixMe
+	}
+}
+
+export interface DistributionsPublicationsListResponse {
+	object?: 'list'
```
Comment (darekg11) [inline packages/sdk/src/types/Distributions.ts:113]: It's always there so please drop `?`

File: packages/sdk/src/types/Distributions.ts [hunk 28]
```diff
@@ -0,0 +1,161 @@
...
+	total?: number
```
Comment (darekg11) [inline packages/sdk/src/types/Distributions.ts:114]: Always there, please drop `?`

File: packages/sdk/src/types/Distributions.ts [hunk 29]
```diff
@@ -0,0 +1,161 @@
...
+	data_ref?: 'publications'
```
Comment (darekg11) [inline packages/sdk/src/types/Distributions.ts:115]: Please drop `?`

File: packages/sdk/src/types/Distributions.ts [hunk 30]
```diff
@@ -0,0 +1,161 @@
...
+	publications?: {
```
Comment (darekg11) [inline packages/sdk/src/types/Distributions.ts:116]: Maybe dfine type for it and do type[]

File: packages/sdk/src/types/Distributions.ts [hunk 31]
```diff
@@ -0,0 +1,161 @@
...
+		id?: string
```
Comment (darekg11) [inline packages/sdk/src/types/Distributions.ts:117]: Please drop `?`

File: packages/sdk/src/types/Distributions.ts [hunk 32]
```diff
@@ -0,0 +1,161 @@
...
+		source_id?: string
+		object?: 'publication'
```
Comment (darekg11) [inline packages/sdk/src/types/Distributions.ts:119]: Please drop `?`

File: packages/sdk/src/types/Distributions.ts [hunk 33]
```diff
@@ -0,0 +1,161 @@
...
+		created_at?: string
```
Comment (darekg11) [inline packages/sdk/src/types/Distributions.ts:120]: Please drop `?`

File: packages/sdk/src/types/Distributions.ts [hunk 34]
```diff
@@ -0,0 +1,161 @@
...
+		customer_id?: string
+		tracking_id?: string
+		metadata?: Record<string, any>
+		channel?: string
+		result?: string
```
Comment (darekg11) [inline packages/sdk/src/types/Distributions.ts:125]: Please drop `?` and it's either `'SUCCESS' | 'FAILURE'`

**Review by darekg11 — CHANGES_REQUESTED** [review]
Changes requested, sorry @pannga I overlook some type definitions in previous review.

File: .changeset/breezy-fans-scream.md [hunk 35]
```diff
@@ -0,0 +1,5 @@
+---
+'@voucherify/sdk': patch
+---
+
+Used FixMe convention for unknown types
```
Comment (eddyw) [inline .changeset/breezy-fans-scream.md:5]: You don't need to create a changeset for every single change 😅 
If changes are to same package and all bump `patch`, you could just edit one of the changesets you've created in this PR:
```suggestion
---
'@voucherify/sdk': patch
---
Distributions types

- Used FixMe convention for unknown types
- Added Distributions types
- Removed deprecated method
- Updated Distribution types
```

It isn't an issue, you could leave it as is if you want (at the end all will be combined by GitHub action) but consider it for next time 😄

File: packages/sdk/src/types/helpers.ts (hunk 26 above)
Comment (pannga) [reply packages/sdk/src/types/helpers.ts:2]: Sorry, I did not stashed it properly

File: packages/sdk/src/types/Distributions.ts [hunk 36]
```diff
@@ -0,0 +1,163 @@
...
+	id: string
+	object: 'publication'
+	created_at: string
+	customer_id: string
+	tracking_id?: string
+	metadata?: Record<string, any>
+	channel?: string
+	source_id?: string
+	result?: 'SUCCESS' | 'FAILURE'
+	customer?: CustomerObject
```
Comment (darekg11) [inline packages/sdk/src/types/Distributions.ts:161]: Please create new type in `Customers.ts` called `SimpleCustomer` - we are using it in couple of places in our API, definition is:
```
id: string
name?: string
email?: string
source_id?: string
metadata?: Record<string, any>
object: "customer"
```
Import it from Customers.ts and use it here

**Review by darekg11 — CHANGES_REQUESTED** [review]
Couple of improvements to be done.

**Review by darekg11 — APPROVED** [review]
Looks good, merging.

**Pull request merged**